### PR TITLE
G3: Support before/after hooks API

### PR DESCRIPTION
**Pull request by denysoblohin-okta**
## Description:

Adds support of [hooks](https://github.com/okta/okta-signin-widget#hooks) for Gen 3.
Hooks can be added with config and added in runtime with `.before` and `.after` methods.

Notes:
1. Emitting `ready` and `afterRender` events [moved](https://github.com/okta/okta-signin-widget/pull/3376/files#diff-d9fe8ef09b73767f05c2fc2b3edc767f2e9db1e1232ae515581b457f66196b0aR386) to useEffect on `widgetRendered` change. This is needed to keep the order of `ready`, `afterRender` events and `after` hooks same as in Gen2
2. `OktaSignIn` class owns instance of `WidgetHooks` class and passes it in widget props
3. There is only 1 discrepancy between in Gen2 and Gen3 in terms of events/hooks:
https://github.com/okta/okta-signin-widget/blob/3fff3b30de02ce50b4f9fd8a1f7ff9f7ecf19d45/test/testcafe/spec/OktaSignIn_spec.js#L604-L610
When user submits form which results in error, context of `afterError` event contains `terminal` form for Gen3, not current form as in Gen2. This is also noticeable in UI: Gen3 hides the form and renders only alert, while Gen2 still presents the form after error.

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Did you add tests, as appropriate, following our [Automated Test guidelines](hxtps://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [x] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-594745](https://oktainc.atlassian.net/browse/OKTA-594745)
- [OKTA-508861](https://oktainc.atlassian.net/browse/OKTA-508861)

### CODE REVIEW

File: src/v3/src/util/widgetHooks.ts
Comment (leemchale-okta) [inline src/v3/src/util/widgetHooks.ts:49]: nice I like this clean way of checking if there are existing hooks and populating them in one line

File: src/v3/src/hooks/useOnSubmit.ts
Comment (leemchale-okta) [inline src/v3/src/hooks/useOnSubmit.ts:257]: does gen 2 have any situations where they would execute the `before` hook before a client transaction?

File: src/v3/src/components/Widget/index.tsx
Comment (leemchale-okta) [inline src/v3/src/components/Widget/index.tsx:397]:
```suggestion
        const isFirstRender = typeof widgetRenderedOnce.current === 'undefined';
```

Comment (leemchale-okta) [reply src/v3/src/components/Widget/index.tsx:397]: nit just to be more specific since its a ref

Comment (leemchale-okta) [reply src/v3/src/components/Widget/index.tsx:397]: or I believe you can just give the ref a default of `false` above so its set to a boolean

File: src/util/Hooks.js
Comment (shawyu-okta) [inline src/util/Hooks.js:34]: Is this all still necessary if we have the default value setting (i.e. `|| []`) below?

File: src/v3/src/components/Widget/index.tsx
Comment (shawyu-okta) [inline src/v3/src/components/Widget/index.tsx:318]: I see you want to pass `transaction` to `handleError` in the `catch` block. Can you use `idxTransaction` instead and eliminate this pattern of having an uninitialized variable?

Comment (shawyu-okta) [reply src/v3/src/components/Widget/index.tsx:318]: Same thing above in `bootstrap`

File: src/v3/src/components/Widget/index.tsx
Comment (shawyu-okta) [inline src/v3/src/components/Widget/index.tsx:209]: Are the `!` non-null assertions needed? I expect the return type of `await authClient.idx.start` would be `idxTransaction` with no possibility of `null | undefined`

File: src/v3/src/util/widgetHooks.ts
Comment (shawyu-okta) [inline src/v3/src/util/widgetHooks.ts:21]: Curious if `hooks` can be kept as a singleton in this file instead of using a `ref` in the `Widget/index.tsx` file and needing to pass the ref around, awkwardly needing to access `ref.current`, etc.

Since it's a `ref` we already are implying it does not affect component rendering so we could potentially simplify by using a singleton. I might also suggest using `Map`s here instead of arrays if we can make that move.

File: src/v3/src/hooks/useOnSubmit.ts
Comment (leemchale-okta) [reply src/v3/src/hooks/useOnSubmit.ts:257]: cc: @glenfannin-okta

File: src/util/Hooks.js
Comment (leemchale-okta) [reply src/util/Hooks.js:34]: I think `hooks[formName]` could contain other hooks other than `before` or `after`?  Which would make this not initialize with default `{ before: [], after: [] };`

Comment (leemchale-okta) [reply src/util/Hooks.js:34]: ignore my last comment, it does look like this is not necessary with the default value setting below

File: src/v3/src/hooks/useOnSubmit.ts
Comment (denysoblohin-okta) [reply src/v3/src/hooks/useOnSubmit.ts:257]: I use `isClientTransaction` as an indicator that request did not succeed. 
In this case "before" and "after" hooks should not be executed for "terminal" form. As it's done for Gen2.

See test `should send "afterError" event (and not execute hooks for "terminal" form) if submit results with error`: https://github.com/okta/okta-signin-widget/pull/3376/files#diff-189986d3f484c4c778c28206906c6fb9e91e4d71404d94ad174eddba3e0acfdcR565

File: src/v3/src/components/Widget/index.tsx
Comment (denysoblohin-okta) [reply src/v3/src/components/Widget/index.tsx]: Actually there is no need to pass transaction to `handleError`. Removed unnecessary code changes.

File: src/util/Hooks.js
Comment (denysoblohin-okta) [reply src/util/Hooks.js]: It's necessary.
I've found a bug in Gen 2 when added new tests for hooks.
If we have only "after" hook in options, and then trying to add new "before" hook with `.before()` API, we'll get `TypeError: Cannot read property 'concat' of undefined`

https://github.com/okta/okta-signin-widget/blob/bc61d536e2aece3e6e1afe2741d25e47295e301f/.widgetrc.sample.js#L20-L25

https://github.com/okta/okta-signin-widget/blob/36fc36650f9a229f0e2ff556c0fea63cead245fc/test/testcafe/spec/OktaSignIn_spec.js#L144-L146

File: src/v3/src/types/widget.ts
Comment (denysoblohin-okta) [inline src/v3/src/types/widget.ts:74]: to allow `class OktaSignIn implements OktaSignInAPI`

File: src/v3/src/util/widgetHooks.ts
Comment (shawyu-okta) [inline src/v3/src/util/widgetHooks.ts:29]: Can we use `this.addHook` instead here? Would be better to use the same public method to add hooks.

I also see the returned type of `hookType` and `hooksByType` is `any` right now so we should address that as well

File: src/v3/src/components/Widget/index.tsx
Comment (shawyu-okta) [inline src/v3/src/components/Widget/index.tsx:218]: Anywhere we are doing a `if(widgetHooks)` guard we can take advantage of optional chaining syntax:
```suggestion
    await widgetHooks?.callHooks('before', undefined, error as Error);
```

Comment (shawyu-okta) [reply src/v3/src/components/Widget/index.tsx:218]: We could also make `widgetHooks` a required prop

File: src/v3/src/components/Widget/index.tsx
Comment (shawyu-okta) [inline src/v3/src/components/Widget/index.tsx:444]: We can set `widgetRenderedOnce` to be `false` by default when creating and avoid needing do this check:
```ts
const widgetRenderedOnce = useRef<boolean>(false);
```
Then we can directly check the value in the conditional below with `if (widgetRenderedOnce.current)`

File: src/v3/src/components/Widget/index.tsx
Comment (shawyu-okta) [inline src/v3/src/components/Widget/index.tsx:445]: Optional chaining simplifies this statement:
```suggestion
        const executeAfterHooks = () => widgetHooks?.callHooks('after', idxTransaction, responseError);
```

File: src/v3/src/hooks/useOnSubmit.ts
Comment (shawyu-okta) [inline src/v3/src/hooks/useOnSubmit.ts:80]: `async` no longer needed

**Review by shawyu-okta — COMMENTED** [review]
Some comments. Thank for you adding testcafe tests! I think you may be to move some of the testcafe spec tests to unit tests for `Hooks.js` to make tests run a bit faster. Either way I would like to see some unit test coverage on `Hooks.js`

File: src/v3/src/util/widgetHooks.ts
Comment (denysoblohin-okta) [reply src/v3/src/util/widgetHooks.ts:53]: I think it's better to not pass `error` at all because its value is not used in method body.
```
callHooks(
    hookType: HookType,
    idxTransaction?: IdxTransaction | null
  )
```
If `idxTransaction` is null, then it's error case and need to execute hooks for "terminal"

File: src/v3/src/types/widget.ts
Comment (denysoblohin-okta) [inline src/v3/src/types/widget.ts:64]: to allow `class OktaSignIn implements OktaSignInAPI`

File: src/v3/src/components/Widget/index.tsx
Comment (denysoblohin-okta) [reply src/v3/src/components/Widget/index.tsx]: I made `widgetHooks` a required prop and pass it to `<Widget>` in [`setup()`](https://github.com/okta/okta-signin-widget/blob/master/src/v3/test/integration/util/setup.tsx#L43) for tests
Otherwise there will be error `Cannot read properties of undefined (reading 'callHooks')`

File: src/v3/src/util/widgetHooks.ts
Comment (shawyu-okta) [inline src/v3/src/util/widgetHooks.ts:30]: TS is making you cast `hookType as HookType` because it can't infer the types of anything that isn't `before` or `after` so it gives you `hooksByType: any`. We both know there will be only `before` or `after` but TS isn't smart enough to know that. Since there's only two cases we could avoid the type cast by directly handling both cases explicitly and writing the inner `for` loop write, once for each case.

Another option is to force TS to understand by using a string array of the literals we care about, which effectively does the same thing but explicitly:
```ts
['before', 'after'].forEach(hookType => {
  for (const hook of (formHooks[hookType] || []) as HookFunction[]) {
    this.addHook(hookType, formName, hook);
  }
});
```

File: src/v3/src/util/widgetHooks.ts
Comment (shawyu-okta) [inline src/v3/src/util/widgetHooks.ts:51]: this length check isn't necessary. if the array is empty the for loop below just won't run

File: src/v3/src/OktaSignIn/index.ts
Comment (shawyu-okta) [inline src/v3/src/OktaSignIn/index.ts:264]: Can you file a ticket to fix this casting? I think it's possible to avoid this but I see why you've done it. Will just take some time with typescript docs and a little bit of creativity but I don't want this one to block the PR. Once you make the jira, please link it in a comment above this type cast.

**Review by shawyu-okta — APPROVED** [review]
Some suggestions I think will be helpful, plus some questions that just need to be answered before merging. Overall lgtm from my side if the notes are addressed.

I'd like to see @lesterchoi-okta take a look over this tomorrow before you merge, especially to ensure that our external `OktaSignIn` API is acceptable since we cannot change that later.

File: src/v3/src/OktaSignIn/index.ts
Comment (denysoblohin-okta) [reply src/v3/src/OktaSignIn/index.ts:180]: Putting `eventEmitter` and `widgetHooks` after `...this.options` is needed to prevent lint error `This spread always overwrites this property.`
I looks fine to put spread at the top as we don't expect overwriting any of listed properties (`authClient`, `globalSuccessFn`, `globalErrorFn `, `eventEmitter` and `widgetHooks`) from options.

File: src/v3/src/components/Widget/index.tsx
Comment (denysoblohin-okta) [reply src/v3/src/components/Widget/index.tsx:447]: Using `async` with `useEffect` will emit type error `'() => Promise<void>' is not assignable to parameter of type 'EffectCallback'`
https://github.com/preactjs/preact/issues/3239
https://github.com/facebook/react/blob/f2c381131fb58c107e6153255bc8d1d6340db506/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L103

File: src/v3/src/components/Widget/index.tsx
Comment (denysoblohin-okta) [reply src/v3/src/components/Widget/index.tsx:452]: Good point. 
I saw your conversation at https://github.com/okta/okta-signin-widget/pull/3290#discussion_r1245414155
Condition `idxTransaction?.status !== IdxStatus.SUCCESS` can be changed to `uischema.elements.length > 0`

File: src/v3/src/OktaSignIn/index.ts
Comment (shawyu-okta) [reply src/v3/src/OktaSignIn/index.ts:180]: Ok, yeah I'm convinced now that in fact we shouldn't spread at the bottom of the object anyways since we don't want to overwrite the `authClient` (or any of the other ones, actually) accidentally since it's always set in the constructor function.

Comment (shawyu-okta) [reply src/v3/src/OktaSignIn/index.ts:180]: Actually we're kind of spreading in extra properties to the `Widget` component we don't need, like `onSuccess`, since it's still in the `options` object after we add the wrappers around them. Not a big deal and no change needed for your PR but realizing we could clean this up a bit.

File: src/v3/src/OktaSignIn/index.ts
Comment (denysoblohin-okta) [reply src/v3/src/OktaSignIn/index.ts:264]: Changed `OktaWidgetEventHandler` declaration to
```
export type OktaWidgetEventHandler = EventCallback | EventCallbackWithError;
```

File: src/v3/src/types/widget.ts
Comment (denysoblohin-okta) [reply src/v3/src/types/widget.ts:114]: As I understand, `WidgetOptions` are "public" options that can be passed to `OktaSignIn` constructor. 
And `WidgetProps` are props passed to `<Widget />`. 
`eventEmitter` and `widgetHooks` are not public, so I moved them to `WidgetProps` (and marked as required as [you suggested](https://github.com/okta/okta-signin-widget/pull/3376#discussion_r1338846208)).

But maybe we can just use
```ts
export type WidgetProps = WidgetOptions;
```
since `Partial` is not needed
and move `eventEmitter` and `widgetHooks`  to `WidgetOptions` as optional. For simplicity.
@lesterchoi-okta ?

Comment (shawyu-okta) [reply src/v3/src/types/widget.ts:114]: Oh you're right, it might be right to keep `WidgetOptions` and `WidgetProps` separate the way you already have them

Comment (shawyu-okta) [reply src/v3/src/types/widget.ts:114]: I wrote a utility type yesterday that might be helpful:
```ts
type PartialExcept<T, K extends keyof T> = Partial<Omit<T, K>> & Pick<T, K>;
```

**Review by lesterchoi-okta — APPROVED** [review]

**Pull request merged**
